### DISCUSSION TITLE
Is the "distributed nodes" feature ready to use?

### DISCUSSION THREAD

**Discussion by shi0rik0** [slawlor/ractor#131] [Q&A] [answered]
I'm looking for a erlang-like framework in rust that supports talking to remote actors via network. In [reddit](https://www.reddit.com/r/rust/comments/113dp70/ractor_not_just_another_actor_framework/) you said that
> Early but stable support for a distributed epmd-like cluster environment, where you can talk to actors over a network link. It's an additional crate (ractor_cluster) that builds on ractor to facilitate the inter-connection between nodes and support remote casts and calls to actors on a remote node.

But the `README.md` says
> On our roadmap is to add more of the Erlang functionality including potentially a distributed actor cluster.
> 
> cluster, which exposes various functionality required for ractor_cluster to set up and manage a cluster of actors over a network link. This is work-in-progress and is being tracked in https://github.com/slawlor/ractor/issues/16.

So I'm wondering whether this cluster feature is ready to use. What's the development stage of this feature?

**Comment by shi0rik0** [comment]
You said that
> Additionally ractor has a companion library, ractor_cluster which is needed for ractor to be deployed in a distributed (cluster-like) scenario. ractor_cluster shouldn't be considered production ready, but it is relatively stable and we'd love your feedback!

Could you tell me why "ractor_cluster shouldn't be considered production ready"? Thanks.

**Reply by slawlor** [reply, marked as answer]
It depends on your definition of production, I have early stage infra that is running it successfully and our integration tests cover a suite of scenarios, however it's not say serving traffic with hundreds of thousands of users.

We haven't yet had the ability to push it to absolute scale limits and see how it performs, so if you're looking at a small scale infrastructure I think it's probably fine to use, there may be some edge bugs that have not yet been found but we believe we've covered most of it. If you're planning on turning it up to a large scale infrastructure, then you may find performance bottlenecks that we have yet to identify. Please feel free to open any issues or missing functionality in your view!

**Reply by shi0rik0** [reply]
Thanks for your great work! Do you know if there are other rust libraries that already provide mature high-performance large-scale clustering support so we can make a comparison?

**Reply by slawlor** [reply]
I mean there's probably something out there, but I don't have any direct experience with anything personally. It's going to depend on what you want out of it, do you also want a distributed **actor** framework? Probably I'd look at something like [Bastion](https://github.com/bastion-rs/bastion) but I haven't used it directly. Additionally maybe actix but they'd deprecated use of their own actor runtime in their web framework.
